File: 1331-rank-transform-of-an-array/1331-rank-transform-of-an-array.java
class Solution {
    public int[] arrayRankTransform(int[] arr) {
        int n = arr.length;
        HashMap<Integer, Integer> map = new HashMap<>();
        int[] sortedArr = new int[n];
        for(int i=0;i<n;i++){
            sortedArr[i] = arr[i];
        }
        Arrays.sort(sortedArr);
        int rank =1;
        for(int i=0;i<n;i++){
            if(i > 0 && sortedArr[i] > sortedArr[i-1]) rank++;
            map.put(sortedArr[i],rank);
        }
        int[]  ans = new int[n];
        for(int i=0;i<n;i++){
            ans[i] = map.get(arr[i]);
        }
        return ans;
    }
}

File: 1331-rank-transform-of-an-array/1331-rank-transform-of-an-array.cpp
class Solution {
public:
    vector<int> arrayRankTransform(vector<int>& arr) {
        int n = arr.size();
        vector<int> sortArr = arr;
        map<int,int> map;
        sort(sortArr.begin(), sortArr.end());
        int rank =1;
        vector<int> ans(n);
        for(int i=0;i<n;i++){
            if(i > 0 && sortArr[i] > sortArr[i-1]){
                rank +=1;
            }
            map[sortArr[i]] = rank;
        }
        for(int i =0;i<n;i++){
            ans[i] = map[arr[i]];
        }
        return ans;
    }
};